{"id":113,"status":"Fixed","summary":"Compiler gives false error with respect to unreachable code","labels":["Type-Defect","Priority-Medium"],"stars":1,"commentCount":5,"comments":[{"id":0,"commenterId":-2020703049423688535,"content":"Try compiling the following in the Closure Compiler UI:\r\n\r\n// \u003d\u003dClosureCompiler\u003d\u003d\r\n// @compilation_level SIMPLE_OPTIMIZATIONS\r\n// @output_file_name default.js\r\n// \u003d\u003d/ClosureCompiler\u003d\u003d\r\n\r\nfunction instanceOf(value, type) {\r\n  try {\r\n    // first try built-in test -- if it succeeds, we\u0027re golden.\r\n    if (value instanceof type) {\r\n      return true;\r\n    }\r\n  } catch (exception) {\r\n    if (exception instanceof TypeError) {\r\n      throw exception; // indicates that \u0026quot;type\u0026quot; is not a type\r\n    }\r\n    // Otherwise, assume the exception was caused by \r\n    // the Firefox 1.0.3 bug.  Work around it.\r\n    return (type \u003d\u003d\u003d Object);\r\n  }\r\n}\r\n\r\nThe Compiler issues the following warning:\r\n\r\nJSC_UNREACHABLE_CODE: unreachable code at line 7 character 0\r\n  } catch (exception) {\r\n\r\nThis code is from a Firefox extension (Chickenfoot) where (at least\r\nhistorically) calling instanceof in this manner could throw a security\r\nexception (or something else, I forget what -- Chickenfoot has been around\r\nsince Firefox 1.0) which is why the catch blocks is there and is indeed\r\nreachable.","timestamp":1266690206,"attachments":[]},{"id":1,"commenterId":-2020703049423688535,"content":"If anyone is in a similar situation, I was able to do the following as a workaround:\r\n\r\n// \u003d\u003dClosureCompiler\u003d\u003d\r\n// @compilation_level SIMPLE_OPTIMIZATIONS\r\n// @output_file_name default.js\r\n// \u003d\u003d/ClosureCompiler\u003d\u003d\r\n\r\nfunction instanceOf(value, type) {\r\n  var result;\r\n  try {\r\n    // first try built-in test -- if it succeeds, we\u0027re golden.\r\n    result \u003d value instanceof type;\r\n  } catch (exception) {\r\n    if (exception instanceof TypeError) {\r\n      throw exception; // indicates that \"type\" is not a type\r\n    }\r\n    // Otherwise, assume the exception was caused by \r\n    // the Firefox 1.0.3 bug.  Work around it.\r\n    return (type \u003d\u003d\u003d Object);\r\n  }\r\n  return result;\r\n}\r\n","timestamp":1266690608,"attachments":[]},{"id":2,"commenterId":-7699928860083865744,"content":"Alan, you should be able to fix this pretty easily. Mind taking a look?","timestamp":1266779512,"attachments":[]},{"id":3,"commenterId":1948758734812428220,"content":"Fixed and checked in.\r\n\r\nThe fix should be available in the next release.","timestamp":1266879682,"attachments":[]},{"id":4,"commenterId":-7699928860083865744,"content":"This issue was closed by revision r114.","timestamp":1266887270,"attachments":[]}]}